spec.md

```diff
@@ -0,0 +1,94 @@
+
+One `HC_GW_ALLOWED_FNS_{app-id}` variable must be set per allowed app id. For example `HC_GW_ALLOWED_FNS_mewsfeed=<zome function list>`.
+
+The variable `HC_GW_ALLOWED_FNS_{app-id}` should permit `*` for users who don't wish to restrict access to the apps on the target Holochain conductor. Note that the gateway is doing nothing else to restrict access to functions that do write data, so opting out of this mechanism is not recommended.
+
+## Request processing
+
+### Validate the request
+
+On receiving a request, the request should be checked:
+
+- The DNA hash string must decode to a valid DNA hash.
+- The coordinator identifier and zome name must be valid UTF-8 and at most 100 characters long.
+- The function name must be a valid function identifier (valid UTF-8 is good enough for now) and at most 100 characters long.
+- The function name must be included in the list of allowed functions for the app.
+- The payload length must be within the configured limit.
+- The payload must be valid base64 url and decode to valid JSON.
+
+### Identify the app to call
+
+The gateway uses a single admin websocket to make requests to Holochain. At this point, either use the existing connection or open a new one if the websocket is closed.
+
+When receiving a valid request, the gateway first checks its cached list of installed apps. If the cache is empty or the app isn't found in the cache, it calls the `ListApps` operation on Holochain with the filter set to only return running apps. The response is stored in the cache for future requests.
+
+The gateway identifies apps that match both the DNA hash specified in the request and have an installed app ID matching the `coordinator-identifier`. It then checks if the app ID is in the list of allowed apps configured in `AllowedAppIds`. If multiple matching apps are found, an error is returned since a unique app cannot be determined.
+
+If no app is found in the initial check but the cache was populated, a fresh request to `ListApps` is made to update the cache, and the search is performed again. If the app still isn't found, or if it's found but not in the allowed list, an appropriate error is returned.
+
+### Prepare zome call signing
+
+At this point, we appear to have received a valid request that identifies an app running on Holochain. We should now try to issue signing credentials for ourselves.
+
+The HTTP gateway uses ClientAgentSigner to manage signing credentials for zome calls, instead of using `sodoken` directly. The gateway then uses its admin API connection to authorize these credentials for each cell via authorize_signing_credentials. The granted functions are set according to the value of HC_GW_ALLOWED_FNS_{app-id}, either as All or a specific list of functions.
```

Comment:
```suggestion
The HTTP gateway uses ClientAgentSigner to manage signing credentials for zome calls, instead of connecting to Holochain's keystore directly. The gateway then uses its admin API connection to authorize these credentials for each cell via authorize_signing_credentials. The granted functions are set according to the value of HC_GW_ALLOWED_FNS_{app-id}, either as All or a specific list of functions.
```